### PR TITLE
Sidecar Rewrite

### PR DESCRIPTION
**Pull request by NohaIhab**
- Wrote 2 Pebble layers for the 2 containers
- Patched the manager and http ports with KubernetesServicePatch
- Added yaml for roles
- Applied resources with k8s resource handler
- Pushed namespace-labels file to profiles container with Container.push
- Updated unit tests
- Added http health checks to the pebble layers
- Added on_remove event handler

to see health for profiles checks run `kubectl exec kubeflow-profiles-0 -n <namespace> -c <container> -- /charm/bin/pebble checks`, container can be kubeflow-profiles or kubeflow-kfam

### CODE REVIEW

File: src/charm.py
Comment (i-chvets) [inline src/charm.py:4]: Add doc string describing what this charm is.

File: src/charm.py
Comment (i-chvets) [inline src/charm.py:81]: Add doc string.

File: src/charm.py
Comment (i-chvets) [inline src/charm.py:29]: Add doc string.

File: src/charm.py
Comment (i-chvets) [inline src/charm.py:26]: Add doc string describing what this charm is.

File: src/charm.py
Comment (i-chvets) [inline src/charm.py:89]: Add doc string.

File: src/charm.py
Comment (i-chvets) [inline src/charm.py:101]: Add doc string.

File: src/charm.py
Comment (i-chvets) [inline src/charm.py:166]: Add doc string.

File: src/charm.py
Comment (i-chvets) [inline src/charm.py:183]: Should we define filenames via constants?
I.e. NAMESPACE_LABELS_FILE="src/files/namespace-labels.yaml"

File: src/charm.py
Comment (i-chvets) [inline src/charm.py:237]: We should change this doc string to something that describe of what is done on in this method.

File: src/charm.py
Comment (i-chvets) [inline src/charm.py:250]: We should change this doc string to something that describe of what is done on in this method.

File: src/charm.py
Comment (i-chvets) [inline src/charm.py:263]: Add doc string.

File: src/charm.py
Comment (i-chvets) [inline src/charm.py:273]: Add doc string. This code is not touched, but we should be bringing it up to standard.

File: src/charm.py
Comment (i-chvets) [inline src/charm.py:282]: Add doc string. This code is not touched, but we should be bringing it up to standard.

File: src/charm.py
Comment (i-chvets) [inline src/charm.py:302]: Add doc string. This code is not touched, but we should be bringing it up to standard.

File: tests/unit/test_operator.py
Comment (i-chvets) [inline tests/unit/test_operator.py:6]: Add doc string describing of what test is doing.

File: tests/unit/test_operator.py
Comment (i-chvets) [inline tests/unit/test_operator.py:17]: Add doc string describing of what test is doing.

File: tests/unit/test_operator.py
Comment (i-chvets) [inline tests/unit/test_operator.py:28]: Add doc string describing of what test is doing.

File: tests/unit/test_operator.py
Comment (i-chvets) [inline tests/unit/test_operator.py:39]: Add doc string describing of what test is doing.

File: tests/unit/test_operator.py
Comment (i-chvets) [inline tests/unit/test_operator.py:69]: Add doc string describing of what test is doing.

File: tests/unit/test_operator.py
Comment (i-chvets) [inline tests/unit/test_operator.py:94]: Add doc string describing of what test is doing.

**Review by i-chvets — CHANGES_REQUESTED** [review]
Added some comments. Most of them are minor style updates.

**Comment by i-chvets** [comment]
Kudos for getting health checks implemented!
 
General comments:
- Remove .coverage file
- Remove .flake8 file
- Maybe rename `files/` directory to `templates/` to be consistent with other charms (also need to update code to reflect that).
- Need to add test to verify that health checks are working (integration).
- Use commit messages format described in coding standards section 2.2.1, i.e. using `feat: <description>` style.

A suggestion:
- Integration tests are somewhat limited. How complex is a deployment? Can we add some basic integration tests to verify functionality of deployed charm?

**Review by i-chvets — APPROVED** [review]

File: .github/workflows/integrate.yaml
Comment (ca-scribner) [inline .github/workflows/integrate.yaml:83]: This needs to be removed before final approval

File: requirements.in
Comment (ca-scribner) [inline requirements.in:8]:
```suggestion
charmed-kubeflow-chisme

```

File: src/templates/auth_manifests.yaml.j2
Comment (ca-scribner) [inline src/templates/auth_manifests.yaml.j2:15]:
```suggestion
      - "*"

```

File: tests/integration/test_charm.py
Comment (ca-scribner) [inline tests/integration/test_charm.py:65]: I can't decide if these add good testing value or if they're more than is needed.  I think they do add value

File: tests/unit/conftest.py
Comment (ca-scribner) [inline tests/unit/conftest.py:33]: There is nothing wrong with this, but imo these are clearer if they're in the test_operator.py file rather than conftest.py.  This is just because anything in conftest.py is a bit hidden from view.  I don't think we have a team convention on where something like this would go though so changes aren't necessary if you prefer this way

File: src/charm.py
Comment (ca-scribner) [inline src/charm.py:194]: This can move outside the `with` block (the `with` is only needed to open and read `labels_file`.

File: src/charm.py
Comment (ca-scribner) [inline src/charm.py:276]: Is raising here the desired action?  If we raise here, the charm will go into an error state.  Is the preference here that we hit an error state (and therefore the application is not removed), or should it *try* to remove these items but if it fails it logs the issues but proceeds anyway?

There's no right or wrong, just making sure it is doing the one you prefer.  Sometimes, especially in a remove hook, we want to *try* to clean up what we can but we're ok with failing and continuing on

File: tests/unit/conftest.py
Comment (NohaIhab) [reply tests/unit/conftest.py:33]: I think it's cleaner for the setup to be separated from the tests, although I see your point on clearance. perhaps add a docstring at the top of the test_operator.py to refer to conftest.py?

File: src/charm.py
Comment (NohaIhab) [reply src/charm.py:194]: good point

File: src/charm.py
Comment (NohaIhab) [reply src/charm.py:276]: that depends on the consequences of not cleaning up the resources, would that affect another deploy per say?
also I think we should reevaluate this same behavior in other charms, dashboard and seldon that I know of.

File: tests/unit/conftest.py
Comment (ca-scribner) [reply tests/unit/conftest.py]: If you're looking for cleanliness, what about storing them in a separate python file (helpers.py, fixtures,py, test_operator_fixtures.py, etc) and importing them from there?  The trouble with `conftest.py` is that it is an *automatically loaded* file.  It is essentially like an inherited global variable.  So now any other test executed from `/tests/unit` or any of it's subdirectories will always inherit your fixtures whether they want them or not.  

I'd also suggest that the lost clarity is worse than gained cleanliness.  Readers of `test_operator.py` no longer need to sift through these fixtures, but they're also not told where to find them if they want to know what they do.  They need to know to look in a conftest.py file and search for them.  Whereas if the fixtures are explicitly imported (like other package imports), there's a clear indicator about what's loaded and where to look at the source code

File: src/charm.py
Comment (ca-scribner) [reply src/charm.py:276]: Yeah this feels like a good discussion in a tech talk, and then us adopting some general guidelines.  I think we've gotten stung by errors in `remove` hooks before, because they block `juju` from removing all the other stuff that it should (eg: juju wont delete all the rest of the things it should until the remove hook is resolved), which iirc caused bigger problems.  But I might be remembering this wrong

File: tests/unit/conftest.py
Comment (NohaIhab) [reply tests/unit/conftest.py]: great explanation, will move to test_operator_fixtures.py and import explicitly

File: src/charm.py
Comment (NohaIhab) [reply src/charm.py:276]: added to tech topics

**Review by i-chvets — APPROVED** [review]
Approved.

File: src/charm.py
Comment (ca-scribner) [reply src/charm.py:281]: After discussions with Juju and internally, it was decided that the best practice is to:
* attempt to delete things, catching any errors
* if we fail to delete for a known and acceptable reason (for example, object no longer exists), then warn in logs but do not set Error state
* if we fail for some odd reason (some unexpected error), that might warrant setting Error

The main reason for this is that setting a charm to an error state in a Remove hook is a hard thing for users to navigate out of, and often is worse than the errors we catch

**Pull request merged**
